File: 백준/Gold/17471. 게리맨더링/게리맨더링.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {

    private static int N;
    private static int[] people;
    private static List<List<Integer>> graph;
    private static List<Integer> aPowerSet;
    private static List<Integer> bPowerSet;
    private static int minDifPeople;

    public static void main(String[] args) throws IOException {
        init();
        findPowerset(1);
        System.out.println(minDifPeople == Integer.MAX_VALUE ? -1 : minDifPeople);
    }

    private static void init() throws IOException {
        BufferedReader input = new BufferedReader(new InputStreamReader(System.in));
        N = Integer.parseInt(input.readLine());
        people = new int[N + 1];
        graph = new ArrayList<>();
        aPowerSet = new ArrayList<>();
        bPowerSet = new ArrayList<>();
        minDifPeople = Integer.MAX_VALUE;

        for (int i = 0; i < N + 1; i++) {
            List<Integer> adjList = new ArrayList<>();
            graph.add(adjList);
        }

        StringTokenizer tokenizer = new StringTokenizer(input.readLine());
        for (int i = 1; i < people.length; i++) {
            people[i] = Integer.parseInt(tokenizer.nextToken());
        }

        for (int i = 1; i < N + 1; i++) {
            tokenizer = new StringTokenizer(input.readLine());
            int n = Integer.parseInt(tokenizer.nextToken());
            for (int j = 0; j < n; j++) {
                int v = Integer.parseInt(tokenizer.nextToken());
                graph.get(i).add(v);
                graph.get(v).add(i);
            }
        }
    }

    private static void findPowerset(int start) {
        if (start == N + 1) {
            makeBPowerSer();
            if (aPowerSet.size() != 0 && aPowerSet.size() != N && isValidArea()) {
                int diff = findDiffPeople();
                minDifPeople = Math.min(minDifPeople, diff);
            }

            return;
        }

        aPowerSet.add(start);
        findPowerset(start + 1);
        aPowerSet.remove(aPowerSet.indexOf(start));
        findPowerset(start + 1);
    }

    private static void makeBPowerSer() {
        bPowerSet.clear();
        for (int i = 1; i <= N; i++) {
            if (aPowerSet.contains(i)) {
                continue;
            }

            bPowerSet.add(i);
        }
    }

    private static int findDiffPeople() {
        int aArea = 0;

        for (int num : aPowerSet) {
            aArea += people[num];
        }

        int bArea = totalPeople() - aArea;

        return Math.abs(aArea - bArea);
    }

    private static int totalPeople() {
        int sum = 0;

        for (int i = 1; i <= N; i++) {
            sum += people[i];
        }

        return sum;
    }

    private static boolean isValidArea() {
        boolean valid = false;
        /// powerSet 검사
        int aCnt = bfs(aPowerSet);
        /// 나머지 검사
        int bCnt = bfs(bPowerSet);
        /// 합쳐서 N과 같으면 valid = true;
        if (aCnt + bCnt == N) {
            valid = true;
        }

        return valid;
    }

    private static int bfs(List<Integer> powerSet) {
        int cnt = 0;
        boolean[] visited = new boolean[N + 1];
        Queue<Integer> que = new LinkedList<>();
        que.offer(powerSet.get(0));
        visited[powerSet.get(0)] = true;

        while (!que.isEmpty()) {
            int cur = que.poll();
            cnt++;

            for (int nextCur : graph.get(cur)) {
                if (!visited[nextCur] && powerSet.contains(nextCur)) {
                    que.offer(nextCur);
                    visited[nextCur] = true;
                }
            }
        }

        return cnt;
    }
}

File: 백준/Gold/17471. 게리맨더링/게리맨더링.cc
#include <bits/stdc++.h>
using namespace std;
int n, m;
int a[11];
int mn = INT_MAX;
vector<int> adj[11];
// 선택한거 vs 선택 안한거 비교하면 되지 않을까?
// 그리고 선택한거끼리 다 연결되어있어야하고 안선태된것도 모두 연결되어있어야하는거 아닌가?
// 일단 0 ~ 1<<n - 1 이렇게 하면 9개 선택 / n개 선택이니까 1 ~ 1<<n - 2해야 하는거 아닌가?
// 맞는거 같음 일단은 
// 그러면 mask (특정 몇개 켜져 있는거) ~mask하면 나머지를 켤 수 있지 않나?
// 그리고나서 모두 연결되어있는지 체크하고 되어있으면 합 반환
// 그리고나서 diff min 갱신 아닌가
// 연결 여부는 어떻게 알지? -> 두개로 나뉘니까 같은 영역인 어딘가와는 반드시 연결되어있어야함
bool valid(int mask) {
    int start = -1;
    for (int i = 0; i < n; i++) {
        if (mask & (1 << i)) {
            start = i; break;
        }
    }

    queue<int> q;
    int vis = 0;
    vis |= (1 << start);
    q.push(start);

    int cnt = 0;
    while (q.size()) {
        cnt++;
        int cur = q.front(); q.pop();
        for (int nxt : adj[cur]) {
            if (!(mask & (1 << nxt))) continue;
            if (vis & (1 << nxt)) {
                continue;
            }
            vis |= (1 << nxt);
            q.push(nxt);
        }
    }
    int need = 0;
    for (int i = 0; i < n; i++) if (mask & (1 << i)) need++;

    return cnt == need;
}
int people(int mask) {
    int sum = 0;
    for(int i = 0; i < n; i++){
        if (mask & (1 << i)) sum+=a[i];
    }

    return sum;
}
int main() {
    cin >> n;
    for (int i = 0; i < n; i++) {
        cin >> a[i];
    }
    for (int i = 0; i < n; i++){
        cin >> m;
        for (int j = 0; j < m; j++){
            int b;
            cin >> b;
            adj[i].push_back(b - 1);
        }
    }

    for (int i = 1; i <= (1 << n) - 2; i++) {
        int amask = i;
        int bmask = ~i;
        for (int j = 0; j < n; j++) {
            if (valid(amask) && valid(bmask)) {
                int diff = abs(people(amask) - people(bmask));
                mn = min(diff, mn);
            }
        }
    }
    if (mn == INT_MAX) {
        cout << -1 << '\n';
        return 0;
    }
    cout << mn << '\n';
    return 0;
}